overlord/state/taskrunner.go

```diff
@@ -73,13 +85,15 @@ func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 
 		r.state.Lock()
 		defer r.state.Unlock()
+		halted := task.HaltTasks()
 		if err == nil {
 			task.SetStatus(DoneStatus)
 		} else {
-			task.SetStatus(ErrorStatus)
+			taskFail(task)
+		}
+		if len(halted) > 0 {
+			r.state.EnsureBefore(0)
```

Comment:
Sorry, don't understand. The logic here is saying "if there are halted tasks, we should fire Ensure so that these may be handled". In the error case above, though, we've marked all of these tasks into a final state saying "won't do".

This is likely a bit optimistic, and perhaps even totally wrong. But perhaps okay for now until we know better. That said, the logic even at it is now seems in conflict. It's pretending to know we should mark tasks in error ahead of time, pretending to know that we should only ensure if tasks are halted, but then pretending it doesn't know that it just did that.